service/frontend/workflow_handler.go

```diff
@@ -1305,13 +1305,30 @@ func (wh *WorkflowHandler) RecordActivityTaskHeartbeatById(ctx context.Context,
 	runID := request.GetRunId() // runID is optional so can be empty
 	activityID := request.GetActivityId()
 
-	if workflowID == "" {
-		return nil, errWorkflowIDNotSet
-	}
 	if activityID == "" {
 		return nil, errActivityIDNotSet
 	}
 
+	// If workflowID is empty, it means the activity is a standalone activity and we need to set the component ref.
+	// Else this should be a validation error.
+	var componentRef []byte
+	if workflowID == "" {
+		if !wh.IsStandaloneActivityEnabled(request.GetNamespace()) {
+			return nil, errWorkflowIDNotSet
```

Comment:
I think this should be `serviceerror.NewUnavailable(activity.StandaloneActivityDisabledError)`
for consistency with `RecordActivityTaskHeartbeat`.